### PR TITLE
XSLT: Upcoming changes

### PR DESCRIPTION
**Pull request by nrwahl2**
For @kgaillot 's perusal

### CODE REVIEW

File: tools/crm_resource_runtime.c [hunk 1]
```diff
@@ -1331,7 +1331,7 @@ update_scheduler_input(pcmk_scheduler_t *scheduler, xmlNode **xml)
         scheduler->input = *xml;
         scheduler->priv->now = crm_time_new(NULL);
     }
-    return pcmk_rc_ok;
+    return rc;
```
Comment (kgaillot) [inline tools/crm_resource_runtime.c:1334]: We'll need this one for 2.1.9-rc2, do a separate PR for this (listing as regression in 2.1.8) then backport it

Comment (nrwahl2) [reply tools/crm_resource_runtime.c:1334]: Noted. This wasn't even supposed to be in this branch. I had it as a precursor to trying to fix another crm_resource bug.

File: doc/sphinx/Pacemaker_Explained/rules.rst [hunk 2]
```diff
@@ -59,9 +59,18 @@ Each context that supports rules may contain a single ``rule`` element.
...
        * ``or``: the rule is satisfied if any condition is satisfied
+   * - .. _negate:
+
+       .. index::
+          pair: rule; negate
+
+       negate
+     - :ref:`boolean <boolean>`
+     - false
+     - If true, evaluate the rule's negation instead of the rule itself
```
[the remark this answers is not in the document]
Comment (nrwahl2) [reply doc/sphinx/Pacemaker_Explained/rules.rst:73]: I really don't understand what's confusing about it. The effect of "negate" is "if the rule would evaluate to true, it evaluates to false instead; if the rule would evaluate to false, it evaluates to true."

[the remark this answers is not in the document]
Comment (nrwahl2) [reply doc/sphinx/Pacemaker_Explained/rules.rst:73]: We evaluate it and THEN flip the result if negate=true

File: xml/nvset-4.0.rng [hunk 3]
```diff
@@ -89,7 +101,12 @@
       <zeroOrMore>
         <element name="nvpair">
           <choice>
-            <attribute name="id-ref"><data type="IDREF"/></attribute>
+            <group>
+              <attribute name="id-ref"><data type="IDREF"/></attribute>
+              <optional>
+                <attribute name="name"><text/></attribute>
+              </optional>
+            </group>
```
[the remark this answers is not in the document]
Comment (nrwahl2) [reply xml/nvset-4.0.rng:144]: Whoops, thanks!

File: xml/nvset-4.0.rng [hunk 4]
```diff
@@ -15,9 +15,21 @@
...
+   RelaxNG does not allow an <except> to contain a <ref>, so these must be
+   included explicitly anywhere the list of unsupported values is overridden.
+   -->
   <define name="element-nvset.value">
     <attribute name="value">
-      <text/>
+      <data type="string">
+        <except>
+          <value>#default</value>
+        </except>
```
[the remark this answers is not in the document]
Comment (nrwahl2) [reply xml/nvset-4.0.rng:31]: I'll do an inventory of places where it is treated specially soon, and we can go from there. It wouldn't be terribly complicated to disallow it only for `meta_attributes` nvpairs

File: doc/sphinx/Pacemaker_Explained/rules.rst (hunk 2 above)
[the remark this answers is not in the document]
Comment (kgaillot) [reply doc/sphinx/Pacemaker_Explained/rules.rst:73]: Right, but trying to express the interaction of the attributes makes the issue more obvious:

* boolean-op:
  * and: If negate is false (the default), the rule is satisfied only if all conditions are satisfied; if negate is true, the rule is satisfied only if all conditions are not simultaneously satisfied;
  * or: If negate is false (the default), the rule is satisfied if any condition is satisfied; if negate is true, the rule is satisfied only if none of the conditions are satisfied

It's even more confusing when you consider that node attribute expressions can use not-equals. It's not confusing at a logical level, it's confusing at the level of a sysadmin trying to quickly follow the configuration. Pacemaker's complexity already daunts most people, so anything that makes it more so has to have a significant justification.

[the remark this answers is not in the document]
Comment (nrwahl2) [reply doc/sphinx/Pacemaker_Explained/rules.rst:73]: The case-by-case interaction of the attributes seems unnecessary and confusing to discuss. Document `negate` something like: "Evaluate the rule and return the negation of the result." Or "If true, Pacemaker first evaluates the rule as if `negate` were absent. Then if that result is true, the rule evaluates to false; if that result is false, the rule evaluates to true."

You're better at concisely and clearly documenting behavior than I am, but you get the idea. `negate` can be thought of as a layer on top of the rest.

IMO it's not even confusing at a sysadmin level. Viewing it as a sysadmin, I was *shocked* that there was no `"not"` operator already built into rules. This is something I would expect to exist already, as basic logical functionality.

The immediate justification is for these XSL transforms of course, but I could see it being useful for negating an existing rule via a wrapper around a nested `id-ref` rule.

[the remark this answers is not in the document]
Comment (kgaillot) [reply doc/sphinx/Pacemaker_Explained/rules.rst:73]: The issue isn't the documentation but being able to grok the configuration easily. How long does it take you to confidently determine the exact situation this defines?

    <rule id="rule1" score="INFINITY" boolean-op="and" negate="true">
        <expression id="expression1" attribute="department" operation="ne" value="IT"/>
        <date_expression id="date1" operation="date_spec">
            <date_spec id="date_spec1" hours="9-17" weekdays="1-5"/>
        </date_expression>
    </rule>

The average sysadmin (who doesn't deal with boolean logic very often) will be in a worse position. Now define the same situation without negate and compare. And the difference in difficulty widens with every expression added to the rule.

[the remark this answers is not in the document]
Comment (nrwahl2) [reply doc/sphinx/Pacemaker_Explained/rules.rst:73]: > How long

About 5 seconds if I ignore the details of the nested expressions, about 45 seconds if I don't

[the remark this answers is not in the document]
Comment (nrwahl2) [reply doc/sphinx/Pacemaker_Explained/rules.rst:73]: If you think the above is that confusing, then I'm sure other people would too. I still fail to see the confusion, but I'm a sample size of 1.

[the remark this answers is not in the document]
Comment (kgaillot) [reply doc/sphinx/Pacemaker_Explained/rules.rst:73]: Yes the details matter to the sysadmin trying to figure out what this does to a critical service :)

With a simple `or` and the inverse conditions instead of `negate`, it's still just a few seconds to grasp the details. That's a big difference when we're talking about a few lines of a large configuration being looked at -- even more so if someone who's already overworked is trying to troubleshoot a cluster issue with management breathing down their necks.

Basically I want to force users to define rules using a positive sense so it's easier to follow for the next person.
